Nullable.md

```diff
@@ -0,0 +1,442 @@
+currently have: in particular, a union type is not a concrete type and, as
+a consequence, there are no values of any non-trivial union type. The use of
+a non-concrete type is important both for optimization purposes and for
+clarifying the possible semantics we can define for the new nullable types,
+as we would be replacing a concrete type with a non-concrete type.
+
+Deciding between these three options will require balancing a variety of
+concerns. The arguments for each option are summarized below. We include all
+of them for clarity, although we believe that the arguments on behalf of
+`Union{Some{T}, Void}` are the strongest.
+
+### `Union{T, Void}` Pros & Cons
+
+When a function returns `Union{T, Void}`, the user can immediately apply all
+methods defined on `T` when the value happens to be of type `T`. Allowing users
+to call existing methods without having to unwrap non-null values can be
+convenient.
+
+In this regard, the use of a `Union{T, NAtype}` pattern in DataArrays was
+perceived by many to offer greater convenience than our current `Nullable{T}`
+type provides, although the use of a union type was also the cause of that
+library's historical poor performance because of the compiler's non-optimized
+handling of union types. This Julep assumes that union types will be optimized
+such that poor performance is no longer a consequence of using a union type as
+the implementation of `?T`.
+
+Instead of concerns about inferior performance, one major argument against
+using `Union{T, Void}` is that a `Union{T, Void}` type can only be used
+reliably in code if the methods being called on instances of the union type
+are defined for both the `T` and `Void` sub-types of that union. Going forward,
+this document will refer to these two cases as the non-null and null cases,
+respectively.
+
+Because the `Union{T, Void}` representation uses a union type, it is possible
+for a method to be defined only over a subset of the values in the union. In
+particular, the possibility of directly calling `f(x::T)` on a value of type
+`Union{T, Void}` means that it is possible for code to handle the non-null case
+without being capable of handling the null case. We will refer to situations in
+which code handles only one of the non-null or null cases without handling the
+other case as failing to exhibit totality, a term used because the code does
+not define an execution path for the totality of all possible inputs.
+
+To ensure totality, users must either (a) use functions for which both
+`f(x::T)` and `f(x::Void)` are defined or (b) use higher-order functions
+that can take in a function with a method of the form `f(x::T)` and employ a
+default semantics for null values to handle the `f(x::Void)` case. In
+particular, we will refer to the pattern of extending an `f(x::T)` method to
+handle `f(x::Union{T, Void)` by assuming that `f(x::Void) = nothing` as
+lifting.
+
+The unique concern with `Union{T, Void}` is `f(x::T)` will often be defined
```

Comment:
I wouldn't say it's the unique concern. The strongest concern for me is that this nullable type wouldn't allow distinguishing between no value and a stored null value when indexing a dictionary.